project.clj

```diff
@@ -9,10 +9,12 @@
                  [com.nedap.staffing-solutions/utils.collections "2.1.0"]
                  [com.nedap.staffing-solutions/utils.modular "2.2.0-alpha3"]
                  [com.nedap.staffing-solutions/utils.spec.predicates "1.1.0"]
+                 [io.reflectoring.diffparser/diffparser "1.4"]
```

Comment:
We use the diff parser to show where the formatter would've wanted to make the change, so we require parsing the diff for meaningful reporting. I will see if I can change the dep to `com.googlecode.java-diff-utils` (already a transient dep)

Reply:
```patch
diff --git a/src/formatting_stack/util.clj b/src/formatting_stack/util.clj
index 8a7df6c..da362f5 100644
--- a/src/formatting_stack/util.clj
+++ b/src/formatting_stack/util.clj
@@ -1,6 +1,6 @@
 (ns formatting-stack.util
   (:require
-   [clojure.java.data :as data]
+   [clojure.java.data :refer [from-java]]
    [clojure.java.io :as io]
    [clojure.spec.alpha :as spec]
    [clojure.tools.namespace.file :as file]
@@ -10,11 +10,13 @@
    [nedap.speced.def :as speced]
    [nedap.utils.collections.eager :refer [partitioning-pmap]]
    [nedap.utils.collections.seq :refer [distribute-evenly-by]]
-   [nedap.utils.spec.predicates :refer [present-string?]])
+   [nedap.utils.spec.predicates :refer [present-string?]]
+   [clojure.string :as str]
+   [clojure.string :as string])
   (:import
    (clojure.lang IBlockingDeref IPending)
-   (io.reflectoring.diffparser.api UnifiedDiffParser)
-   (java.io File)))
+   (java.io File)
+   (difflib DiffUtils)))
 
 (defmacro rcomp
   "Like `comp`, but in reverse order.
@@ -144,33 +146,39 @@
 (defn colorize [s color]
   (str \u001b (ansi-colors color) s \u001b (ansi-colors :reset)))
 
-(spec/def ::grouping-spec
-  (spec/cat ::neutral (spec/* (comp #{"NEUTRAL"} :lineType))
-            ::from    (spec/+ (comp #{"FROM"} :lineType))
-            ::rest    (spec/* any?)))
-
 (speced/defn ^coll? diff->line-numbers
-  "Returns maps of consecutive removals/changes (as `:begin` to `:end`) per `:filename` in `diff`."
+  "Returns maps of consecutive changes (as `:begin` to `:end`) per `:filename` in `diff`."
   [^string? diff]
-  (->> (io/input-stream (.getBytes diff))
-       (.parse (UnifiedDiffParser.))
-       (data/from-java)
-       (mapcat (fn [{:keys [toFileName hunks]}]
-                 (->> (loop [{::keys [from rest]}
-                             (->> hunks
-                                  (mapcat (fn [{:keys [lines] {:keys [lineStart]} :fromFileRange}]
-                                            (->> lines
-                                                 (remove (comp #{"TO"} :lineType))
-                                                 (map-indexed (fn [idx line] (assoc line :lineNumber (+ idx lineStart)))))))
-                                  (spec/conform ::grouping-spec)) ;; group consecutive FROM-changes
-                             result []]
-                        (if (seq rest)
-                          (recur (spec/conform ::grouping-spec rest)
-                                 (conj result from))
-                          (if from
-                            (conj result from)
-                            result)))
-                      (map (fn [grouped-lines]
-                             {:filename toFileName
-                              :begin (apply min (map :lineNumber grouped-lines))
-                              :end (apply max (map :lineNumber grouped-lines))})))))))
+  (->> (string/split diff #"diff --git ")
+       (drop 1) ;; first match is empty
+       (mapcat (speced/fn [^present-string? file-diff]
+                 (speced/let [[_, ^present-string? filename] (re-find #"^a\/(.*) b\/" file-diff)]
+                   (->> (str/split-lines file-diff)
+                        (DiffUtils/parseUnifiedDiff)
+                        (from-java)
+                        (:deltas)
+                        (mapcat (speced/fn [{{^coll? rev :lines}                       :revised
+                                             {:keys [^coll? lines ^pos-int? position]} :original}]
+                                  (let [no-additions (->> rev
+                                                          (filter (fn [idx x]
+                                                                    (= x (get (vec lines) idx)))))]
+                                    (println ::no-additions no-additions)
+                                    (->> lines
+                                         (keep-indexed (fn [idx x]
+                                                         (when-not (= (get (vec no-additions) idx) x)
+                                                           (println ::xxx x)
+                                                           (println ::rev (get (vec no-additions) idx))
+                                                           (+ 1 position idx))))
+                                         (reduce (fn [acc line-number]
+                                                   (let [{:keys [begin end]
+                                                          :or   {begin 0, end 0}} (last acc)]
+                                                     (if (= (inc end) line-number)
+                                                       ;; replace last one if consecutive
+                                                       (conj (butlast acc) {:begin    begin
+                                                                            :end      line-number
+                                                                            :filename filename})
+                                                       ;; add new entry otherwise
+                                                       (conj acc {:begin    line-number
+                                                                  :end      line-number
+                                                                  :filename filename}))))
+                                           [])))))))))))
```

This impl has a bug where it can't detect moved code, because I compare the diff-parts myself instead of the parser.

----

The diffutils-parser can only deal with per-file diffs, so we need to partially parse the diff ourselves in this case. 

Keeping that in mind, and seeing how small and effective the diffparser is, so we'll need to keep it in. We can re-evaluate this choice in the future as it's not public api anyway. Will add a couple more tests.

I found a way around the `::grouping-spec` this way (so this wasn't a total waste of efforts).